### ISSUE TITLE
What is the source of randomness  in get_thetas?

### ISSUE THREAD

**Issue by martinritchie** [MilaNLProc/contextualized-topic-models#21]
I am experimenting with your code and trying to understand the source of randomness in [`get_thetas`](https://github.com/MilaNLProc/contextualized-topic-models/blob/master/contextualized_topic_models/models/ctm.py#L238), why are multiple sampled needed to be averaged? 

I have the model set in evaluation mode and fixed my seeds, if I change the seeds I get different  topic  predictions. I am sorry if this is a rudimentary question but I can't figure out  the source of this randomness.

Thank you, 

Martin

**Comment by vinid** [comment]
Hi @martinritchie!

the randomness comes from the fact that the distribution is sampled from the [network](https://openreview.net/pdf?id=BybtVK9lg), that is why we need to sample multiple times to get a more reliable estimate. If we do not do the average, our distribution is based on a single sample and it can be thus not reliable as a distribution.

The first time we ran our [experiments](https://github.com/MilaNLProc/contextualized-topic-models/blob/master/examples/multilingual-topic-modeling.ipynb) we acutally got bad results because we weren't averaging the distribution. The more samples you get the more reliable the predictions are. That is also why I'd not use a seed, if you use a seed for the entire PyTorch network you are going to always sample the same distribution (that might be a bad one).

Hope this helps :)

**Comment by martinritchie** [comment]
Hi, 

Thank you for your reply. 

My gut feeling was along those lines, however for inference: the network is in eval mode (dropout is off) and weights are not getting updated. So I can't understand why changing seeds produces such drastically different results: changing the seed results in the network converging on different thetas.  

Given the above and since I am looking use this approach elsewhere I must ensure repeatability. I am happy to fix the seed but I  wish to understand why changing the seed results in convergence to different thetas. 

Thanks again,

Martin

**Comment by vinid** [comment]
Could you share something with me to try to replicate your experiments? It's weird that the average always converges to different numbers.

Can you tell me which seeds are you setting and how many samples are you taking for the average?

**Comment by martinritchie** [comment]
Just to be clear, if I use  the same seed for inference as I do during training I have no problems and it always converges to the same result. However, if I use a different  seed for inference it does converge (and always to the same value) but to a different value compared to the training seed and this results in a different classification. 

To experiment with this yourself (unfortunately I can not share the data) I set the seeds to different values and infer topics for the same passage of text. 

My gut feeling is that there is nothing wrong, I just can't pin down the source of randomness. 

Thank you for your replies. If I figure it out before you i'll post my results before you (its always possible I have made a mistake somewhere).
